src/io/reader.rs

```diff
@@ -0,0 +1,209 @@
+            format: Some(format),
+        }
+    }
+
+    /// Get the currently determined format.
+    pub fn format(&self) -> Option<ImageFormat> {
+        self.format
+    }
+
+    /// Supply the format as which to interpret the read image.
+    pub fn set_format(&mut self, format: ImageFormat) {
+        self.format = Some(format);
+    }
+
+    /// Remove the current information on the image format.
+    ///
+    /// Note that many operations require format information to be present and will return e.g. an
+    /// `ImageError::UnsupportedError` when the image format has not been set.
+    pub fn clear_format(&mut self) {
+        self.format = None;
+    }
+
+    /// Unwrap the reader.
+    pub fn into_inner(self) -> R {
+        self.inner
+    }
+}
+
+impl Reader<BufReader<File>> {
+    /// Open a file to read, format will be guessed from content or path.
+    ///
+    /// This will automatically begin reading the start of the file to determine the format, then
+    /// fall back to the path. Use [`from_path`] to open a file but guess its format solely based
+    /// on the path and not perform any io on the file yet.
+    ///
+    /// It is always possible to redetect the current format with [`guess_format`].
+    ///
+    /// [`from_path`]: #method.from_path
+    /// [`guess_format`]: #method.guess_format
+    pub fn open<P>(path: P) -> io::Result<Self> where P: AsRef<Path> {
+        Self::open_impl(path.as_ref())
+    }
+
+    /// Open a file to read, format will be guessed from path.
+    ///
+    /// This will not attempt any io operation on the opened file.
+    ///
+    /// If you want to inspect the content for a better guess on the format, which does not depend
+    /// on file extensions, use [`open`] instead.
+    ///
+    /// [`open`]: #method.open
+    pub fn from_path<P>(path: P) -> io::Result<Self> where P: AsRef<Path> {
+        Self::from_path_impl(path.as_ref())
+    }
+
+    fn open_impl(path: &Path) -> io::Result<Self> {
+        let file = File::open(path)?;
+
+        let mut reader = Reader {
+            inner: BufReader::new(file),
+            format: None,
+        };
+
+        reader.guess_format()?;
+
+        if reader.format.is_none() {
+            // If content didn't success, maybe guess the format based on path.
+            if let Ok(format) = ImageFormat::from_path(path) {
+                reader.format = Some(format);
+            }
+        }
+
+        Ok(reader)
+    }
+
+    fn from_path_impl(path: &Path) -> io::Result<Self> {
+        let file = File::open(path)?;
+        Ok(Reader {
+            inner: BufReader::new(file),
+            format: ImageFormat::from_path(path).ok(),
+        })
+    }
+}
+
+impl<R: BufRead + Seek> Reader<R> {
+    /// Replace the format with a guess based on the content.
+    ///
+    /// If the guess fails, the format is unchanged. Error is for seek, hence io::Error
+    /// and not ImageError.
+    pub fn guess_format(&mut self) -> io::Result<()> {
```

Comment:
My personal preference would be to have a single function:
* `with_guessed_formatk(&mut self) -> io::Result<Self>`

In the normal case users could one of:
* `let r = image::Reader::from_path(...)?;` or
* `let r = image::Reader::from_path(...)?.with_guessed_format()?;`

And if they had some obscure requirements there is still the option to have complete flexibility via something fancy like:
```rust
let mut r = image::Reader::from_path(...)?;
let ext = r.format();
r.clear_format();
r.with_guessed_format()?;
let guessed = r.format();
r.set_format(/* some decision based on `ext` and `guessed` */)
```

Reply:
Methods of the style `fn(self) -> Result<Self>` have two disadvantages that make awkward to use as inherent methods of builders with io, that are not `Copy`. This applies to all methods that take `self` by value.

* They can not be applied behind a (mutable) reference. If some component should configure the reader but not consume the stream then it should not require being passed the potentially large value. It should be possible to operate on a reference here.
* The error case with return type `io::Result<Self>` always drops the inner reader. This makes it more complicated and expensive than necessary to use a reader with cleanup such as `BufReader`.

I'd prefer having methods that can operate on `self` by reference which would be simpler than an artifical interface that allows this despite the value semantics on everything else.

Reply:
Those cases don't seem super common so instead of duplicating every method, perhaps we could just make a method `as_mut(&mut self) -> &mut R`? This would allow a lot of the same flexibility: For any `Reader<R>` you can make a `Reader<&mut R>` that wouldn't drop/consume the original, and you do also do anything else you wanted to the underlying reader prior to decoding.

Reply:
I don't see the point exposing a mutable reference to reader if there is no utility around seeking to the prior position (which needs to exist anyways since `guess_format` requires this). I can live with dropping the method that only deduces the format but doesn't set it but a method taking `fn set_guessed_format(&mut self)` instead of `with_guess_format(self)` would be nearly as easy to use:

```
let mut r = image::Reader::from_path(...)?;
r.set_guessed_format()?;
r
```

Reply:
Well, except maybe `by_ref(&mut self) -> Reader<&mut R>` which also preserves the deduced metadata.

Reply:
If the underlying reader supported seek then you'd be able to use it directly. If not, then guess_format wasn't going to be available anyway.

Regardless, I'd like to see this land more than I care about the exact interface. Is there anything else blocking merging

Reply:
The decision of whether to have `into_decoder` right now or simply have `decode` and provide a `DynamicDecoder` (w.o.r.t. bikeshedding) later. Since the current free functions provide a way to query the image dimension metadata without having to construct a decoder I'd rather have an `into_dimensions` similar to that as well (and maybe for the other methods of decoder as well). We can always soft deprecate them later.

Reply:
@fintelia Alright, I've been swayed. The main purpose is to expose a reader and not as an uncommon way of getting `guess_format` to work. So let's do `with_guessed_format(self) -> io::Result<Self>` and see where it takes us. Can always add the other methods later, as above there is enough room in the naming scheme.

Reply:
Lets hold off on `into_decoder` for now, I think that could use a dedicated PR for discussion. It seems fine to have some into metadata functions now and deprecate them once we get proper dynamic decoder support